### PR TITLE
Rebuiliding baseline tree

### PR DESCRIPTION
**Pull request by MarekM25**
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18362.1139 (1903/May2019Update/19H1)
Intel Core i5-8250U CPU 1.60GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.201
  [Host]     : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT
  Job-GYEDKG : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT

Runtime=.NET Core 3.1  IterationCount=10  

```
|                                  Method | NumberOfLeaves |             Mean |            Error |           StdDev |       Gen 0 |     Gen 1 |     Gen 2 |     Allocated |
|---------------------------------------- |--------------- |-----------------:|-----------------:|-----------------:|------------:|----------:|----------:|--------------:|
|             **BuildTreeWithInstantHashing** |             **32** |      **1,258.52 μs** |         **4.285 μs** |         **2.834 μs** |    **162.1094** |         **-** |         **-** |     **502.41 KB** |
|            BuildTreeWithHashingInTheEnd |             32 |         99.70 μs |         0.678 μs |         0.355 μs |     24.9023 |         - |         - |      76.48 KB |
| InsertingValuesWithoutCalculatingHashes |             32 |         11.50 μs |         0.089 μs |         0.047 μs |      5.7983 |         - |         - |      17.77 KB |
|             **BuildTreeWithInstantHashing** |           **1000** |     **43,182.81 μs** |       **186.289 μs** |       **123.219 μs** |   **4583.3333** |  **333.3333** |   **83.3333** |   **15511.23 KB** |
|            BuildTreeWithHashingInTheEnd |           1000 |      2,162.98 μs |        24.997 μs |        14.875 μs |    277.3438 |  113.2813 |         - |    1295.39 KB |
| InsertingValuesWithoutCalculatingHashes |           1000 |        374.06 μs |        36.986 μs |        24.464 μs |    105.9570 |   31.7383 |         - |     408.09 KB |
|             **BuildTreeWithInstantHashing** |         **100000** | **26,310,462.87 μs** | **1,306,820.677 μs** |   **864,380.965 μs** | **465000.0000** | **9000.0000** | **2000.0000** | **1586447.72 KB** |
|            BuildTreeWithHashingInTheEnd |         100000 |  7,414,494.44 μs | 2,247,245.670 μs | 1,337,299.852 μs |  20000.0000 | 5000.0000 | 1000.0000 |  132955.33 KB |
| InsertingValuesWithoutCalculatingHashes |         100000 |    754,170.92 μs |    92,205.884 μs |    54,870.243 μs |   4000.0000 | 1000.0000 |         - |   31187.29 KB |

### CODE REVIEW

File: src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs [hunk 1]
```diff
@@ -193,6 +195,29 @@ public BaselineTreeNode[] GetLeaves(params uint[] leafIndexes)
             return leaves;
         }
 
+        public void RecalculateHashes()
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs:198]: you will need  method that will take the start leaf index too

**Comment by MarekM25** [comment]
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18362.1139 (1903/May2019Update/19H1)
Intel Core i5-8250U CPU 1.60GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.201
  [Host]     : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT
  Job-SRECNI : .NET Core 3.1.3 (CoreCLR 4.700.20.11803, CoreFX 4.700.20.12001), X64 RyuJIT

Runtime=.NET Core 3.1  IterationCount=1  WarmupCount=1  

```
|                       Method | NumberOfLeaves |    Mean | Error |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|----------------------------- |--------------- |--------:|------:|------------:|-----------:|----------:|----------:|
| BuildTreeWithHashingInTheEnd |        1000000 | 702.7 s |    NA | 300000.0000 | 53000.0000 | 4000.0000 |   1.19 GB |

File: src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs (hunk 1 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs:198]: > you will need method that will take the start leaf index too

I know :) I planned it in next PRs :)

Comment (MarekM25) [reply src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs]: @tkstanczak added :)

File: src/Nethermind/Nethermind.Baseline.Benchmark/BuildingTreeBenchmarks.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        [Benchmark]
+        public void BuildTreeWithHashingInTheEnd()
+        {
+            BaselineTree baselineTree = new ShaBaselineTree(new MemDb(), Array.Empty<byte>(), 0);
+            for (uint i = 0; i < _testLeaves.Length; ++i)
+            {
+                baselineTree.Insert(_testLeaves[i], false);
+            }
+
+            baselineTree.CalculateHashes();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Baseline.Benchmark/BuildingTreeBenchmarks.cs:67]: Can you change it to something like:

using (var batch = baseLineTree.StartBatch())
{
  batch.Insert(_testLeaves[i]);
  // or baselineTree.Insert
}

and that would calculate hashes at the end, makes API better IMO - similar like when you batch things to DB

File: src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs [hunk 3]
```diff
@@ -193,6 +202,32 @@ public BaselineTreeNode[] GetLeaves(params uint[] leafIndexes)
...
+                for (uint indexAtRow = startingRowIndex; indexAtRow < nodesToIterate; indexAtRow += 2)
+                {
+                    Index index = new Index(row, indexAtRow);
+                    var hash = LoadValue(index);
+                    var siblingIndex = index.Sibling();
+                    var siblingHash = LoadValue(siblingIndex);
+                    byte[] parentHash = new byte[32];
+                    Hash(hash.Bytes.AsSpan(), siblingHash.Bytes.AsSpan(), parentHash);
+                    Index parentIndex = index.Parent();
+                    SaveValue(parentIndex, parentHash);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs:220]: maybe you can get significant speed increase if you batch insert into the database

File: src/Nethermind/Nethermind.Baseline/BaselineModule.cs [hunk 4]
```diff
@@ -526,26 +480,27 @@ private BaselineTree RebuildEntireTree(Address treeAddress, Keccak blockHash)
...
+            using var batch = baselineTree.StartBatch();
             foreach (FilterLog filterLog in insertLeavesLogs
                 .Union(insertLeafLogs)
                 .OrderBy(fl => fl.BlockNumber).ThenBy(fl => fl.LogIndex))
             {
                 if (filterLog.Data.Length == 96)
                 {
                     Keccak leafHash = new Keccak(filterLog.Data.Slice(32, 32).ToArray());
-                    baselineTree.Insert(leafHash);
+                    baselineTree.Insert(leafHash, false);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Baseline/BaselineModule.cs:491]: Lukasz suggested inserts on batch -> this way you could avoid loading the data from the DB when recalculating hashes
You currently do

Insert0, CommitDb, Insert1, CommitDb, Insert2, CommitDb, Insert3, CommitDb
Load0, Load1, Insert01, CommitDB, Load2, Load3, Insert23 CommitDb
Load01, Load23, Insert0123, CommitDb

What you could do instead would be
Insert0, Insert1, Insert2, Insert3
Insert01, Insert23
Insert0123,
CommitDb

**Review by tkstanczak — CHANGES_REQUESTED** [review]
See the idea with in-flight hash clculation (you can do that by intercepting the Load operation so it reads from the current batch first, and only then from DB)

**Pull request closed without merging**